WebSocket Support (SS over WSS) and OutlineCaddy Integration

lunarthegrey opened the pull request:
This PR adds WebSocket transport support for Shadowsocks connections, enabling SS over WSS to bypass restrictive network filters. It fixes https://github.com/Jigsaw-Code/outline-server/issues/1676 where much of the work was discussed.

## Features

### WebSocket Transport Support
- New listener-based API model allowing access keys to specify WebSocket listeners alongside traditional Shadowsocks
- WebSocket server automatically enabled when a WebSocket access key is created
- `GET /access-keys/{id}` returns YAML configuration for WebSocket-enabled keys (Outline Client v1.15.0+)
- Supports both TCP over WebSocket (`websocket-stream`) and UDP over WebSocket (`websocket-packet`)

### Embedded OutlineCaddy Server
- New [outline_caddy_server.ts](cci:7://file:///Users/lunar/Documents/git/outline-server/src/shadowbox/server/outline_caddy_server.ts:0:0-0:0) module for managing Caddy as the WebSocket reverse proxy
- YAML-based Caddyfile configuration with automatic HTTPS via ACME
- API proxy support allowing the management API to be accessed via TLS-terminated Caddy
- Internal API proxy uses HTTPS with self-signed cert verification skipped

---

## API Changes

### New Endpoints
- `PUT /server/listeners-for-new-access-keys` - Configure listener types for new keys
- `PUT /server/web-server` - Configure embedded Caddy web server

### Modified Endpoints
- `GET /access-keys/{id}` - Returns JSON for traditional keys, **YAML for WebSocket keys**
- `POST /access-keys` - Now accepts `listeners` array parameter

### AccessKey Schema
- Added optional `listeners` field (`tcp`, `udp`, `websocket-stream`, `websocket-packet`)
- Added optional `dynamicConfig` field for WebSocket transport configuration
- `password`, `port`, `method`, `accessUrl` are now optional (omitted for WSS-only keys)

---

## Build & CI Changes

### Docker Build Workflow
- Rewrote multi-arch build to use Task-based builds instead of Docker buildx matrix
- Added Node.js, Go, and Task setup steps
- Builds both `amd64` and `arm64` images sequentially
- Creates and pushes multi-arch manifest

### GitHub Actions Concurrency
- Added workflow-specific prefixes to concurrency groups to prevent cross-workflow cancellation
  - `build-and-test-*` for build workflow
  - `license-*` for license checks

### Taskfile Changes
- Added `download_xcaddy` task for cross-platform xcaddy downloads
- Build task now includes OutlineCaddy binary built via xcaddy with plugins:
  - `outlinecaddy@v0.0.1`
  - `caddy_yaml_adapter`
  - `caddy-l4@v0.0.0-20251201210923-0c96591f5650`

---

## Dependencies

### Go Dependencies (`go.mod`, `go.sum`)
- Updated `prometheus/client_golang` to v1.20.5
- Updated `prometheus/common` to v0.62.0
- Updated `oschwald/geoip2-golang` to v1.11.0
- Updated `golang.org/x/crypto` to v0.32.0
- Updated `golang.org/x/sync` to v0.11.0
- Updated `google.golang.org/protobuf` to v1.36.4
- Bumped outline-ss-server to v1.9.2

---

## Documentation

### README.md
- Added "WebSocket Support (SS over WSS)" section with usage examples
- Documented listener configuration API
- Added YAML response example for WebSocket keys

### api.yml
- Full OpenAPI documentation for new endpoints
- Updated AccessKey schema with optional SS fields
- Documented dual response types (JSON/YAML) for `GET /access-keys/{id}`

---

## Usage Examples

```sh
### Configure WebSocket Listeners

curl --insecure -X PUT -H "Content-Type: application/json" \
  -d '{
    "tcp": {"port": 443},
    "udp": {"port": 443},
    "websocketStream": {"path": "/tcp", "webServerPort": 8080},
    "websocketPacket": {"path": "/udp", "webServerPort": 8080}
  }' \
  $API_URL/server/listeners-for-new-access-keys

### Enable Caddy Web Server

curl --insecure -X PUT -H "Content-Type: application/json" \
  -d '{
    "enabled": true,
    "autoHttps": true,
    "email": "admin@example.com",
    "domain": "your-domain.com"
  }' \
  $API_URL/server/web-server

### Create WebSocket-Enabled Access Key

curl --insecure -X POST -H "Content-Type: application/json" \
  -d '{"name": "WSS User", "listeners": ["tcp", "udp", "websocket-stream", "websocket-packet"]}' \
  $API_URL/access-keys

### Get WebSocket Key Config (returns YAML)

curl --insecure $API_URL/access-keys/1

lunarthegrey committed:
Add support for SS over WSS into shadowbox

lunarthegrey committed:
Attempt to fix shadowbox builds

lunarthegrey committed:
Attempt to fix shadowbox builds

lunarthegrey committed:
Attempt to fix shadowbox builds

lunarthegrey committed:
Attempt to fix shadowbox linting errors

lunarthegrey committed:
Attempt to fix license-check

lunarthegrey committed:
Attempt to fix dynamic configs

lunarthegrey committed:
Bump outline-ss-server to v1.9.2

lunarthegrey committed:
Attempt to fix shadowbox builds

lunarthegrey committed:
Attempt to fix shadowbox builds

lunarthegrey committed:
Attempt to fix dynamic access key display issues

lunarthegrey committed:
Attempt to fix dynamic access key display issues

lunarthegrey committed:
Rethink how dynamic access keys are displayed

lunarthegrey committed:
Update .gitignore

lunarthegrey committed:
Remove dynamicAccessKeyUrl as it's not necessary

lunarthegrey committed:
Improve error handling for WebSocket keys

lunarthegrey committed:
Migrate WebSocket support to listener-based API model

lunarthegrey committed:
Satisfy linting

lunarthegrey committed:
Attempt to fix tests

lunarthegrey committed:
Fix yaml for dynamic keys

lunarthegrey committed:
Attempt to fix builds

lunarthegrey committed:
Enable WebSocket server when a WebSocket key is created

lunarthegrey committed:
Improve WS server config insertion when new WS keys are added

lunarthegrey committed:
Add embedded OutlineCaddy integration and listener-aware WebSocket support

lunarthegrey committed:
Fix builds

lunarthegrey committed:
Update releases

lunarthegrey committed:
Update go.sum

lunarthegrey committed:
Attempt different method for outlinecaddy build

lunarthegrey committed:
Fix builds

lunarthegrey committed:
Fix builds

lunarthegrey committed:
Update Taskfile to use `trimPrefix` instead of `replace` for XCADDY version number extraction.

lunarthegrey committed:
Update caddy-l4 plugin version in Taskfile.yml

lunarthegrey committed:
Add whitespace for consistency in manager_service.spec.ts.

lunarthegrey committed:
Include listeners in manager service server info property checks and apply formatting.

lunarthegrey committed:
Correctly assert access key properties when created without listeners

lunarthegrey committed:
Add WebSocket support (SS over WSS) documentation and refactor manager service tests.

lunarthegrey committed:
Add workflow-specific prefixes to GitHub Actions concurrency groups.

lunarthegrey committed:
Add spaces around curly braces in imports and object literals

lunarthegrey committed:
Update manager service tests to correctly await asynchronous calls and apply minor formatting.

lunarthegrey committed:
Switch Caddy configuration from JSON to YAML

lunarthegrey committed:
Add Caddy API proxy support with a new configuration option and Caddyfile template changes

lunarthegrey committed:
Add spaces around curly braces in import statements and object literals.

lunarthegrey committed:
Add dynamic configuration to WebSocket-enabled access keys in the API response.

lunarthegrey committed:
Make Shadowsocks-specific access key fields optional and conditionally include them in the API response based on listener types.

lunarthegrey committed:
Adjust dataLimit field inclusion in access key API responses for all key types.

lunarthegrey committed:
Consolidate dynamic access key config into `/access-keys/{id}` endpoint and remove dedicated dynamic config endpoint.

fortuna commented on src/shadowbox/Taskfile.yml:
```diff
@@ -18,6 +18,32 @@ requires:
   vars: [OUTPUT_BASE]
 
 tasks:
+  download_xcaddy:
```
Can't you just call `go run github.com/caddyserver/xcaddy/cmd/xcaddy`?

The way we do this is to add it as a tool, like we do here:
https://github.com/Jigsaw-Code/outline-sdk/blob/fa686d9463ac56af75aa846f0b446547e73a38b7/go.mod#L53

Then you can simply call `go tool xcaddy`, and we get the proper version pinning in the go.mod/sum

That sounds a lot simpler and removes all this code we need to maintain. Can you do the go tool approach instead? It can be in its own PR to speed up the review/approval process.

lunarthegrey replied:
Good point, will change this.

fortuna commented on src/shadowbox/Taskfile.yml:
```diff
@@ -40,6 +66,16 @@ tasks:
         vars: {TARGET_DIR: '{{.BIN_DIR}}'}
       # Set CGO_ENABLED=0 to force static linkage. See https://mt165.co.uk/blog/static-link-go/.
       - GOOS={{.TARGET_OS}} GOARCH={{.GOARCH}} CGO_ENABLED=0 go build -ldflags='-s -w -X main.version=embedded' -o '{{.BIN_DIR}}/' github.com/Jigsaw-Code/outline-ss-server/cmd/outline-ss-server
+      - task: download_xcaddy
```
Remove the download and call `go tool xcaddy`

fortuna commented on src/shadowbox/Taskfile.yml:
```diff
@@ -40,6 +66,16 @@ tasks:
+          XCADDY_OS: '{{.TARGET_OS}}'
+          XCADDY_ARCH: '{{.GOARCH}}'
+      - |
+        '{{joinPath .BIN_DIR "xcaddy"}}' build --output '{{joinPath .BIN_DIR "outline-caddy"}}' \
```
It's ok to build outlinecaddy for now, but we should be releasing it instead, and using the released binary in the future: https://github.com/Jigsaw-Code/outline-ss-server/releases/tag/outlinecaddy%2Fv0.0.1

I think we haven't established a release process for it yet.

lunarthegrey replied:
Will keep this in mind, thanks.

fortuna reviewed (changes requested):
Let's simplify how we build outilnecaddy so it's easier to reason about and there's less to maintain. I recommend putting it in a standalone PR so it's not blocked on the rest of the PR

fortuna commented on src/shadowbox/server/server_config.ts:
```diff
@@ -41,7 +68,7 @@ export interface ServerConfigJson {
   experimental?: {
     // Whether ASN metric annotation for Prometheus is enabled.
-    asnMetricsEnabled?: boolean;  // DEPRECATED
+    asnMetricsEnabled?: boolean; // DEPRECATED
```
We use two spaces before the comment. Please revert.

fortuna reviewed (changes requested):
I don't think we are ready to accept this PR. We need to understand how this will play with dynamic access keys. With dynamic keys we don't need to set things for "new keys". We can do it for all keys, simplifying the mental model.

We may want a separate API for dynamic keys, so this needs more discussion.

I encourage you to send the build & CI changes as a PR though.

lunarthegrey committed:
Refactor `outline-caddy` build to use `go tool xcaddy` and rename `CaddyWebServerConfig` to `WebServerConfig`

lunarthegrey committed:
Update Go module dependencies

lunarthegrey committed:
refactor(api): make listeners mutable and add dedicated dynamic config endpoint

API Changes:
- Rename PUT /server/listeners-for-new-access-keys → PUT /server/listeners
- Add `applyToExisting` parameter to optionally update all existing keys
- Add PUT /access-keys/{id}/listeners for per-key listener updates
- Add GET /access-keys/{id}/dynamic-config for YAML transport config
- Make GET /access-keys/{id} always return JSON (use dynamic-config for YAML)

Type Renames:
- ListenersForNewAccessKeys → ListenersConfig
- listenersForNewAccessKeys config field → listeners

This simplifies the mental model by:
1. Allowing listener updates on any key, not just at creation
2. Providing a clear global vs per-key control pattern
3. Separating JSON metadata from YAML transport config

lunarthegrey committed:
Merge branch 'OutlineFoundation:master' into master

lunarthegrey commented:
Made a new commit https://github.com/OutlineFoundation/outline-server/pull/1685/commits/be31e617fd6e24f2974959d80dfb412dec32ac05 which (I think) simplifies some of the logic and potentially some of the concerns you've had @fortuna. Let me know if you want to discuss further.

lunarthegrey committed:
Update Caddy server's reverse proxy transport protocol to HTTP for local API communication.

lunarthegrey committed:
Merge branch 'OutlineFoundation:master' into master

rslay commented:
Adding support for Shadowsocks over websocket connections would be a drastic improvement, I would personally love this feature since for quite some time, Shadowsocks traffic has been easy to identify and block in certain regions ([source from gfw.report](https://gfw.report/blog/gfw_shadowsocks/)), but if the traffic looks like it's complex HTTP packets which might be necessary for web functionality (over websockets) it would make identifying and blocking the traffic much more painful for censors!

@fortuna I know the PR is quite large, but there's a lot of people (myself included) that would give this added functionality some use
